### PR TITLE
dns: added DNSForwardParallel to support sending parallel queries to all name servers

### PR DESCRIPTION
**Pull request by fatedier**
**Please provide a description of this PR:**
Fixes: #39598

### CODE REVIEW

File: pkg/dns/client/dns.go [hunk 1]
```diff
@@ -396,15 +402,59 @@ func (h *LocalDNSServer) queryUpstream(upstreamClient *dns.Client, req *dns.Msg,
...
 	return response
 }
 
+// queryUpstreamParallel will send parallel queries to all nameservers and return first response immediately.
+func (h *LocalDNSServer) queryUpstreamParallel(upstreamClient *dns.Client, req *dns.Msg, scope *istiolog.Scope) *dns.Msg {
+	responseCh := make(chan *dns.Msg, 1)
+	errCh := make(chan error, len(h.resolvConfServers))
+
+	queryOne := func(upstream string) {
+		cResponse, _, err := upstreamClient.Exchange(req, upstream)
```
Comment (howardjohn) [inline pkg/dns/client/dns.go:418]: should we use ExchangeContext so we can cancel other requests once one completes?

**Review by howardjohn — COMMENTED** [review]
At a high level seems reasonable. I would like to see some examples of other go implementations (coredns, go builtin, tailscale, etc all have DNS implementations) to make sure we are following any standards, best practices, edge cases, etc here.

In particular, using alpine as an example of why this is good behavior doesn't seem super compelling given how many times i have heard of alpine causing DNS issues. It may still be, just a bit suspicious of using only alpine as a source of motivation here.

File: pkg/dns/client/dns.go (hunk 1 above)
Comment (fatedier) [reply pkg/dns/client/dns.go:418]: I tested it but failed. `ExchangeContext` can't be canceled by context. It just extract deadline from the context.

**Comment by fatedier** [comment]
> At a high level seems reasonable. I would like to see some examples of other go implementations (coredns, go builtin, tailscale, etc all have DNS implementations) to make sure we are following any standards, best practices, edge cases, etc here.
> 
> In particular, using alpine as an example of why this is good behavior doesn't seem super compelling given how many times i have heard of alpine causing DNS issues. It may still be, just a bit suspicious of using only alpine as a source of motivation here.

@howardjohn I agree. Parallel DNS queries are not suitable for all scenarios, so I make it optional and not the default behavior. In my scenario, I configure nameservers the ip of node-local-dns and coredns. When the node-local-dns is down, requests to external service will delay up to 8 seconds and failed.  This is almost unacceptable in microservices and lead to multiple nameservers meaningless.

I searched some documents:

* [dnsmasq](https://thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html#:~:text=etc/resolv.conf-,%2D%2Dall%2Dservers,-By%20default%2C%20when): `--all-servers` Setting this flag forces dnsmasq to send all queries to all available servers. The reply from the server which answers first will be returned to the original requester.
* [tailscale](https://tailscale.com/kb/1194/dns-resolver-order/?q=dns): Tailscale will either proxy all DNS requests—in which case we query all nameservers in parallel and use the quickest response—or defer to the operating system, which we have no control over.
* [RFC 6731](https://datatracker.ietf.org/doc/html/rfc6731.txt#appendix-A.1): It's about RDNSS. I'm not very sure if it's related.
* [fanout](https://coredns.io/explugins/fanout/): A coredns external plugin, parallel proxying DNS messages to upstream resolvers.

By the way, coredns's forward plugin also support [health check](https://coredns.io/plugins/forward/#:~:text=on%20sequential%20ordering.-,health_check,-configure%20the%20behaviour) of the upstream servers. Maybe it's an alternative but don't resolve the problem completely.

File: pilot/cmd/pilot-agent/options/options.go [hunk 2]
```diff
@@ -98,6 +98,9 @@ var (
 	DNSCaptureAddr = env.RegisterStringVar("DNS_PROXY_ADDR", "localhost:15053",
 		"Custom address for the DNS proxy. If it ends with :53 and running as root allows running without iptable DNS capture")
 
+	DNSForwardParallel = env.RegisterBoolVar("ISTIO_META_DNS_FORWARD_PARALLEL", false,
```
Comment (howardjohn) [inline pilot/cmd/pilot-agent/options/options.go:101]:
```suggestion
	DNSForwardParallel = env.RegisterBoolVar("DNS_FORWARD_PARALLEL", false,
```

and clear the model/context.go change

The istio_meta_ and NodeMetadata should only be needed when we are sending this data to Istiod. But this option only impacts the agent locally so no need

**Review by howardjohn — APPROVED** [review]

File: pilot/cmd/pilot-agent/options/options.go (hunk 2 above)
Comment (fatedier) [reply pilot/cmd/pilot-agent/options/options.go:101]: Got it, thanks.

**Pull request merged**